BraveShared/CertificatePinning.swift

```diff
@@ -0,0 +1,160 @@
+    }
+    
+    public func evaluate(_ trust: SecTrust, forHost host: String) throws {
+        // Certificate validation
+        guard !certificates.isEmpty else {
+            throw error(reason: "Empty Certificates")
+        }
+        
+        // Self signed anchoring
+        if options.contains(.allowSelfSigned) {
+            guard SecTrustSetAnchorCertificates(trust, certificates as CFArray) == errSecSuccess else {
+                throw error(reason: "Self Signed Certificate Anchor Failed")
+            }
+            
+            guard SecTrustSetAnchorCertificatesOnly(trust, true) == errSecSuccess else {
+                throw error(reason: "Self Signed Certificate Anchor Only Failed")
+            }
+        }
+        
+        // Default validation
+        if options.contains(.default) {
+            guard SecTrustSetPolicies(trust, SecPolicyCreateSSL(true, nil)) == errSecSuccess else {
+                throw error(reason: "Trust Set Policies Failed")
+            }
+            
+            if #available(iOS 13, *) {
+                var err: CFError?
+                if !SecTrustEvaluateWithError(trust, &err) {
+                    if let err = err as Error? {
+                        throw error(reason: "Trust Evaluation Failed: \(err)")
+                    }
+                    
+                    throw error(reason: "Unable to Evaluate Trust")
+                }
+            } else {
+                var result: SecTrustResultType = .invalid
+                guard SecTrustEvaluate(trust, &result) == errSecSuccess, result == .unspecified || result == .proceed else {
+                    throw error(reason: "Trust Evaluation Failed")
+                }
+            }
+        }
+        
+        // Host validation
+        if options.contains(.validateHost) {
+            guard SecTrustSetPolicies(trust, SecPolicyCreateSSL(true, host as CFString)) == errSecSuccess else {
+                throw error(reason: "Trust Set Policies for Host Failed")
+            }
+            
+            if #available(iOS 13, *) {
+                var err: CFError?
+                if !SecTrustEvaluateWithError(trust, &err) {
+                    if let err = err as Error? {
+                        throw error(reason: "Trust Evaluation Failed: \(err)")
+                    }
+                    
+                    throw error(reason: "Unable to Evaluate Trust")
+                }
+            } else {
+                var result: SecTrustResultType = .invalid
+                guard SecTrustEvaluate(trust, &result) == errSecSuccess, result == .unspecified || result == .proceed else {
+                    throw error(reason: "Trust Evaluation Failed")
+                }
+            }
+        }
+        
+        // Certificate binary matching
+        let serverCertificates = Set((0..<SecTrustGetCertificateCount(trust))
```

Comment:
The certificates in the `Assets` directory are root certs for Global/DigiCert. Wouldn't it be better to fetch the certificate data of the root cert only? `SecTrustGetCertificateAtIndex` should get the root cert for index = max.

Reply:
The code loads ALL the certificates in the chain for "Certificate Chain Validation"..
In my personal experience, I have never seen an app that pins against the ROOT certificate as it compromises security and pretty much defeats the purpose of pinning unless you really trust the CA.

Usually we pin against the Leaf certificate. As we go up the chain of certificates, security decreases. The leaf is the lowest in the chain, then intermediate, then root.

For this reason, I did not know why we were pinning the root but I decided to leave it how it originally was and just validate the chain still. If any of the roots are compromised, pinning should fail.

There's a nice article on pinning here: https://medium.com/@appmattus/android-security-ssl-pinning-1db8acb6621e

It's for Android, but it applies to every system. Specifically the section: `**Which certificate should you pin against in the chain?**`

Reply:
validating the complete chain is a good idea, but the leaf and intermediate certs change frequently and its hard to have a new release with every change. I think the change is good as is.